src/java/org/apache/cassandra/io/sstable/format/SSTableReaderBuilder.java

```diff
@@ -106,6 +107,11 @@ public static FileHandle.Builder defaultIndexHandleBuilder(Descriptor descriptor
         return StorageProvider.instance.fileHandleBuilderFor(descriptor, component);
     }
 
+    public static FileHandle.Builder primaryIndexWriteTimeBuilder(Descriptor descriptor, Component component, OperationType operationType)
```

Comment:
Do you know when things fails without this change, that is, when is the code that opens the primary index files (of the compacted files) before they have been uploaded during compaction (and similar processes)?

Asking because it could be nice to document those places. The context here is that it's kind of not absolutely ideal that we do so and in theory, it's not clear why we need to (we don't need it for flush, so why would we for other process writing sstables?). There reason I'm mentioning all this is that that initially, flush was also doing this (opening the primary index file "too early") and I fixed it with NSS to avoid needing the very method added here, and I kind of wonder how hard fixing it here actually is.

And the context why I fixed it in the first place is that those "open files for read while they are still being written" methods kind of get in the way of writing things directly into remote storage in the future, which I think we ultimately want.

I'll finish adding that the `indexBuildTimeFileHandleBuilderFor` method also gets in the way of such future "direct streaming to remote storage" for indexes, but the reason I added that one is that I didn't knew enough about SAI to know how to fix it so I left that for later. More precisely, in the case of the primary index at flush, there was not real reason why the index was opened "at write time", this was just the code opening the file earlier than it needed too, and that's why I'm wondering if this isn't the case here too. For SAI index however where the code genuinely read one index component in order to finish writing another component, so fixing that is more involved.

Anyway, all that to say that I don't mind terribly if we add this as a shortcut, but if it's not too hard to having needing it in the first place, then that would be my personal preference.

Reply:
Thanks for the context. 

```
INFO  [docker-java-stream--1386328826] 2024-08-14_12:59:46.676 [compactor] STDOUT: org.apache.cassandra.io.sstable.format.trieindex.PartitionIndex.load(PartitionIndex.java:201)
INFO  [docker-java-stream--1386328826] 2024-08-14_12:59:46.676 [compactor] STDOUT: org.apache.cassandra.io.sstable.format.trieindex.PartitionIndex.load(PartitionIndex.java:189)
INFO  [docker-java-stream--1386328826] 2024-08-14_12:59:46.676 [compactor] STDOUT: org.apache.cassandra.io.sstable.format.trieindex.PartitionIndex.load(PartitionIndex.java:178)
INFO  [docker-java-stream--1386328826] 2024-08-14_12:59:46.676 [compactor] STDOUT: org.apache.cassandra.io.sstable.format.trieindex.TrieIndexSSTableWriter$IndexWriter.completedPartitionIndex(TrieIndexSSTableWriter.java:433)
INFO  [docker-java-stream--1386328826] 2024-08-14_12:59:46.676 [compactor] STDOUT: org.apache.cassandra.io.sstable.format.trieindex.TrieIndexSSTableWriter.openFinal(TrieIndexSSTableWriter.java:230)
INFO  [docker-java-stream--1386328826] 2024-08-14_12:59:46.676 [compactor] STDOUT: org.apache.cassandra.io.sstable.format.SortedTableWriter$TransactionalProxy.openResult(SortedTableWriter.java:357)
INFO  [docker-java-stream--1386328826] 2024-08-14_12:59:46.676 [compactor] STDOUT: org.apache.cassandra.io.sstable.format.SortedTableWriter.openResult(SortedTableWriter.java:324)
INFO  [docker-java-stream--1386328826] 2024-08-14_12:59:46.676 [compactor] STDOUT: org.apache.cassandra.io.sstable.SSTableRewriter.doPrepare(SSTableRewriter.java:352)
INFO  [docker-java-stream--1386328826] 2024-08-14_12:59:46.676 [compactor] STDOUT: org.apache.cassandra.utils.concurrent.Transactional$AbstractTransactional.prepareToCommit(Transactional.java:168)
INFO  [docker-java-stream--1386328826] 2024-08-14_12:59:46.676 [compactor] STDOUT: org.apache.cassandra.db.compaction.writers.CompactionAwareWriter.doPrepare(CompactionAwareWriter.java:111)
INFO  [docker-java-stream--1386328826] 2024-08-14_12:59:46.676 [compactor] STDOUT: org.apache.cassandra.db.compaction.unified.ShardedCompactionWriter.doPrepare(ShardedCompactionWriter.java:116)
INFO  [docker-java-stream--1386328826] 2024-08-14_12:59:46.676 [compactor] STDOUT: org.apache.cassandra.utils.concurrent.Transactional$AbstractTransactional.prepareToCommit(Transactional.java:168)
```
This is the stack track that reads the content before upload. I will try to delay it

Reply:
In the case of flush, flushed sstables are added into the tracker via `replaceFlushed`, instead of using the offline flush transaction. We can open flushed sstable after `transaction.prepareToCommit()` where the flush `transaction` doesn't track those flushed sstables.

Compaction writer is different, it expects the `transaction.update(reader, false)` to happen before `transaction.prepareToCommit()`. No change can be made to `transaction` after READY_TO_COMMIT. 

An alternative approach would be to change CNDB upload from operation transaction to per-sstable transaction. WDYT?

Reply:
> An alternative approach would be to change CNDB upload from per-operation-level transaction to per-sstable-writer transaction.

Do you mean to try to make it so upload happens during the `writer.prepareToCommit()` (or something similar)?

If so, I suppose it's an option, but I'm not sure how much work that is. I'm ok to leave that as a "later" thing though if that's too involved (and I find the transaction ordering stuff quite messy in the first place). Thanks for looking though.

Reply:
> Do you mean to try to make it so upload happens during the writer.prepareToCommit() (or something similar)?

right. it requires some works to re-write RemoteLogTransaction in CNDB